WordPress/gutenberg
# Don't apply root padding to nested layout blocks

## What?


Fixes #43095 by resetting each root padding rule so it doesn't apply where there are nested `.has-global-padding` selectors. The new rules added simply negate the existing rules if the `.has-global-padding` block is nested inside another `.has-global-padding` block.

## Behaviour before fix

* Add left/right root padding value to all blocks with `.has-global-padding`;
* Set negative left/right margin with same value on all `.alignfull` direct children of blocks with `.has-global-padding`;
* Set left/right root padding value on all blocks that are direct children of the `.alignfull` blocks with negative margins.


## Expected behaviour with fix applied

1. The outermost container block with content width set will get left/right root padding value;
2. Its direct `alignfull` children will get negative left/right margins;
3. The direct children of those alignfull children will also get left/right root padding value.
4. Any nested blocks with content width set will _not_ get left/right root padding value _unless_ they are also direct children of the `alignfull`blocks with negative margins mentioned in 2..

I've tested this with a bunch of different nesting patterns and it seems to be working as expected. The one place where there can be a bit of weirdness is in the post editor view, because the `.has-global-padding` class isn't output at root level even if the Post Content block has layout set. This means that a block with layout in the post editor might display some extra padding:
<img width="638" alt="Screen Shot 2022-09-05 at 3 54 47 pm" src="https://user-images.githubusercontent.com/8096000/188370255-9e497296-bee5-44c0-a0ad-74e0c7d904a1.png">
which won't appear in the front end view:
<img width="561" alt="Screen Shot 2022-09-05 at 3 55 10 pm" src="https://user-images.githubusercontent.com/8096000/188370301-3168df55-046f-4b34-812a-a7e2b2576b94.png">

I think the only way to fully fix that is to make it possible for the post editor to be aware of the Post Content block settings, which we should look into doing separately from this fix.



## Testing Instructions




1. In theme.json, add `"useRootPaddingAwareAlignments": true` inside the settings object;
2. In both post and site editor, add a few container blocks (e.g. Group) nested inside each other, with other content inside them. Try with both full width and regular width set, and try adding full width content (e.g. Image block) inside them too;
3. Check that root padding behaves as expected.

## Screenshots or screencast 

Before:
<img width="968" alt="Screen Shot 2022-09-05 at 4 30 14 pm" src="https://user-images.githubusercontent.com/8096000/188376977-c72e2b66-cc5b-4cd3-883d-e971296d1746.png">

After:
<img width="989" alt="Screen Shot 2022-09-05 at 4 29 21 pm" src="https://user-images.githubusercontent.com/8096000/188377072-f1a7024a-64fd-44d3-b71b-c0406637fee2.png">
